src/sentry/replays/lib/query.py

```diff
@@ -0,0 +1,152 @@
+"""Dynamic query parsing library."""
+from typing import Any, Generator, List, Optional, Tuple
+
+from snuba_sdk import Column, Condition, Op
+from snuba_sdk.orderby import Direction, OrderBy
+
+from sentry.api.event_search import SearchFilter
+
+# Interface.
+
+
+OPERATOR_MAP = {
+    "=": Op.EQ,
+    "!=": Op.NEQ,
+    ">": Op.GT,
+    ">=": Op.GTE,
+    "<": Op.LT,
+    "<=": Op.LTE,
+}
+
+
+class Field:
+    attribute_name: Optional[str] = None
+
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        field_alias: Optional[str] = None,
+        query_alias: Optional[str] = None,
+        is_filterable: bool = True,
+        is_sortable: bool = True,
+        operators: Optional[list] = None,
+        validators: Optional[list] = None,
+    ) -> None:
+        self.field_alias = field_alias or name
+        self.query_alias = query_alias or name
+        self.is_filterable = is_filterable
+        self.is_sortable = is_sortable
+        self.operators = operators or self._operators
+        self.validators = validators or []
+
+    def deserialize_operator(self, operator: str) -> Tuple[Any, List[str]]:
+        op = OPERATOR_MAP.get(operator)
+        if op is None:
+            return None, ["Operator not found."]
+        elif op not in self.operators:
+            return None, ["Operator not permitted."]
+        else:
+            return op, []
+
+    def deserialize_value(self, value: str) -> Tuple[Any, List[str]]:
+        try:
+            typed_value = self._python_type(value)
+        except ValueError:
+            return None, ["Invalid value specified."]
+
+        for validator in self.validators:
+            error = validator(typed_value)
+            if error:
+                return None, [error]
+
+        return typed_value, []
+
+
+class String(Field):
+    _operators = [Op.EQ, Op.NEQ]
+    _python_type = str
+
+
+class Number(Field):
+    _operators = [Op.EQ, Op.NEQ, Op.GT, Op.GTE, Op.LT, Op.LTE]
+    _python_type = int
+
+
+class Nested(Field):
+    _operators = []
+    _python_type = None
+
+
+class QueryConfig:
+    def __init__(self, only: Optional[Tuple[str]] = None) -> None:
+        self.only = only
+        self.fields = {}
+
+        for field_name in self.only or dir(self):
```

Comment:
instead of using dir you could instead use `self.__class__.__dict__` which would remove a lot of the boilerplate keys this has to filter, but it may just make sense to set the fields in a dict instead of on the class itself in order to avoid any metaprogramming mishaps.

Reply:
This is a good catch and normally I would agree with __dict__ over dir.  However, __dict__ does not lookup keys stored on parent classes.  If you want to inherit configurations it would break in a way that I think would be unexpected to a user.

Its an interesting point to replace the class with a dict like `{"x": StringField()}` or similar.  Ultimately that's what I'm doing internally just with an extra step.  The dir issue from above could be resolved by the dict syntax.  Subclassing could be handled by issuing `my_config.update(other_config)`.  Configs could be initialized with functions rather than inits.

Ultimately I'm choosing the class based design because of its interface.

1. The class based approach conforms to the Serializer and Model behaviors and is easier for users to grok.
2. Inheritance is easier.
3. No global scope issues to worry about (though you could intentionally sabotage yourself).
4. No weird mutability issues.
5. The overhead from dict construction is the same of object construction (single digit microseconds to iterate over `dir`'s output).

Reply:
sounds good! 👍🏼